### ISSUE TITLE
uv_os_gethostname on Windows 7 x64

### ISSUE THREAD

**Issue by kampeador** [libuv/libuv#3260]
Hello.
This commit uses GetHostNameW and makes the whole libuv unusable on Windows 7 x64. It just crashes.
I am using a MinGW compiler.
https://github.com/libuv/libuv/pull/3149

**Comment by vtjnash** [comment]
Microsoft dropped support for Windows 7 last year, so we did too: https://github.com/libuv/libuv/pull/2821

**Comment by kampeador** [comment]
I know about "supported platform list" and expected this Microsoft employee style answer.
EOL -> Kill it.

1. If you look at libuv source code you will see that is designed with Windows Vista and up.
Extra functionality is loaded dynamically.
2. Even _WIN32_WINNT 0x0600 is defined(Windows Vista)
https://github.com/libuv/libuv/blob/v1.x/include/uv/win.h#L23
3. This is an utility function which most software don't even use.
That change provide nothing significant and there should be an alternative solution.

Why drop anything if everything worked fine before?
What the point of maintaining binary compatibility with all that "_ex" mess when all applications will just crash next version anyway?
To summarize your logic:
before:
`runLibuvCode();`
after:
```
if(!Windows8orAbove) { crashAnApplication(); }
else{ runLibuvCode(); }
```

What company is behind this project?
Is this project alive or just "on life support"?
What happend with v2?
Code looks bloated now. There are even parts which handle Windows XP cases. Old platforms getting dropped while the new ones are still underused. No RIO support and other Win8+ stuff(for example).

**Comment by bnoordhuis** [comment]
@kampeador I've marked your comment off-topic because it's borderline trolling, it's clear you trying to get a rise out of @vtjnash. You're free to participate but remember you're a guest here - either you're civil or you're out.

**Comment by kampeador** [comment]
@bnoordhuis 
I am sorry, but I thought that @vtjnash is trolling me, like I am either blind and braindead.
Supported platform list is on the main page.
I've created this issue to hear a logic behind this and got an "Automatic bot style copy paste answer". Thanks Captain Obvious.

Previously libuv dropped platforms if there was a relatively high maintance cost i.e. different code paths like conditional variables on Windows XP etc. That's not the case here.
Besides that, I am assume you will not accept any PRs which will resolve this problem, because this issue was instantly closed.
It means that is not a community driven project, because I've never seen such nonsense. This in turn rised several questions, which were marked as "borderline trolling".
Maybe I should've create a discussion somewhere instead of "issue", but its still an issue and I've planned to submit a PR for it.
According to your general hostility to potential new contributors it will probably won't make much sense.

**Comment by bnoordhuis** [comment]
You can read about the rationale for dropping Windows 7 support in #2821 - there are new APIs in Windows 8 that we want to use without contorting the code for older versions. You may not like that decision but ultimately it's up to us maintainers to make that call.

I won't reject a pull request outright but it's unlikely to get merged unless the changes are strictly trivial and unintrusive.

**Comment by vtjnash** [comment]
I have been trolled worse before. He seemed to ask a question, then answer it himself at the end (why do we drop support? to add new functionality). The point of supported platforms is exactly to avoid case-by-case questions such as this, so that users and maintainers have shared expectations.

I would not be opposed to changing this to use GetProcAddress, since mingw-w64 on Win32 also only will gain for support for this in the next release of GCC (https://github.com/libuv/libuv/issues/3178), so currently I am also not able to use this commit and sadly have it temporarily reverted on my fork.

**Comment by Spudz76** [comment]
Windows 7 is supported by updates until [January 10, 2023](https://docs.microsoft.com/en-us/lifecycle/faq/extended-security-updates) and would expect the same from software.

**Comment by bnoordhuis** [comment]
Extended Support Updates are where you pay Microsoft big money to provide updates. Are you paying us?

**Comment by Spudz76** [comment]
No, but I'm not paying M$ either to be fair, ESU seems like it came with the VLK key I'm on.  There are lots of reasons to keep using Win7 even without updates, like when embedded stuff like ATMs still ran XP until 10 was out, now they run a stripped down Win7 like WinPE.  I don't think Win10 can ever be abused into a portable edition due to too bloated and too much phone-home junk.

There are use-cases for Win7 beyond my own and regardless of ESU that was just reference for an official date when other people expect Win7 to exist until.  I'm literally not running Win10 on mining rigs behind firewall(s) because Win7 still works and wastes less resources.  Libuv is used heavily in miner code.

**Comment by evpobr** [comment]
Please don't. The end of official support does not mean that people will immediately stop using it.

https://gs.statcounter.com/os-version-market-share/windows/desktop/worldwide

https://www.theverge.com/2021/1/6/22217052/microsoft-windows-7-109-million-pcs-usage-stats-analytics

**Comment by Spudz76** [comment]
I suggest the Warzone uses libuv and probably this is the reason Win7 is blocked from working anymore after 9/8

Downhill destruction from this decision...

**Comment by evpobr** [comment]
In fact, a series of errors occurred.

Now I tried to change the host name of my computer to another one, with cyrillic characters, and Windows honestly warned that such a name could cause problems.

But there was also an error in the libuv code. The string returned by the `gethostname` function was converted from UTF-16 to UTF-8 for some reason, although it is in ANSI encoding. In theory, you need a double conversion, from ANSI to UTF-16, and then from UTF-16 to UTF-8.

But this is in theory. In practice, my test code, when trying to get a hostname with Cyrillic, fails with an error at the initialization stage on `WSAStartup`.

As a result, I think the working option may be as follows: try to get the `GetHostNameW` address dynamically (with `GetProcAddress`), if it exists, use it, otherwise, go back to `gethostname` and correctly convert the result to UTF-8,

**Comment by Spudz76** [comment]
xmrig has bypassed the *broken* `uv_os_gethostname` by just using regular getting-the-hostname code for that part.

**Comment by nicolasnoble** [comment]
Potential fix/workaround for mingw at #3340. This will make the code compile properly for current mingw compilers, while still keeping the fix in place, which will now require Windows 8+ at runtime instead, only when calling the `uv_os_gethostname` function.

**Comment by sp-sabik** [comment]
I came here because of Neovim, which also uses libuv, so no longer runs on Win7 x64.

I understand the complexities of developing an open source project, but I think that breaking backward compatibility in this particular case (GetHostNameW) is not reasonable.

**Comment by bnoordhuis** [comment]
Neovim (or you) can float the patch from #3285 if they feel strongly. We're pretty lenient when it comes to supporting obsolete platforms<sup>1</sup> but not when there is a non-trivial amount of code involved, as is the case here.

<sup>1</sup> If you wanted a reason to switch from Windows to Linux: we still support Linux 2.32, released 12 years ago today.

**Comment by sp-sabik** [comment]
@bnoordhuis Thanks for the link.

I switched to Linux a long time ago.
Unfortunately, I have to work with Windows, even with legacy on Win7.

You can treat this obsolete platform as you like, Microsoft also is trying to bury it as quickly as possible, but it will still exist for a long time.

And if people pay attention to this issue/platform, even provide you with patches, then it is not reasonable to reject them purely for ideological reasons, because "We're pretty lenient when it comes to supporting obsolete platforms".

**Comment by vtjnash** [comment]
We currently have no reviewers with Windows 7 to be able to accept that patch. It exists, and seems reasonable to merge, but we need others to evaluate it. 

https://lwn.net/Articles/876228/

**Comment by Spudz76** [comment]
I have Windows 7 on a lot of things (and will never upgrade) so I'd be available for testing and reviewing if it extends support.

Of course I don't use it as a daily driver, that's Linux.  It was my gaming rig but then Warzone eliminated Win7 support (probably because they use libuv behind the scenes, ironically, it happened at a suspiciously similar time...)  So now I don't play anymore, that's fine, I'm not upgrading.

**Comment by Isopod** [comment]
What is the status on this? I'm not familiar with libuv (but am familiar with C and have worked with similar event libraries before), but I'd be willing to help. If the maintainers feel uneasy about the reverse-engineered version, could we at least fallback to the "ansi" version of gethostname as done [here](https://github.com/Isopod/neovim/blob/390b7c590e5e03ef649b02dbc00ff9ffb4c0249f/third-party/patches/libuv-gethostnamew-win7.patch)? Anything is better than a linkage failure.

**Comment by bnoordhuis** [comment]
@Isopod https://github.com/libuv/libuv/issues/3260#issuecomment-890938947 - the patch from #3285 doesn't meet that bar.

**Comment by Isopod** [comment]
@bnoordhuis What bar are you referring to? What is the problem with the patch from #3285? That it is too long? That it's not tested enough?

**Comment by bnoordhuis** [comment]
Both. If you want a rough guideline: diffs bigger than 10-20 lines are unlikely to qualify as trivial and/or unobtrusive.

The mingw32 fix from #3340 did meet the bar and possibly takes the worst sting out for win7 users. Hasn't been released yet though.

**Comment by EricMCornelius** [comment]
Just out of curiosity - how are you claiming to support semver and yet deprecate entire OS platforms in minor releases?

This is a major breaking change for a number of ecosystems - if you want to be resistant to the needs of the community here you should at the very least acknowledge it via a major semver version bump.

**Comment by bnoordhuis** [comment]
Semver doesn't say anything about platform support. Libuv's compatibility promise pertains to its public API and ABI. We reserve the right to drop or demote support for obsolete platforms.

**Comment by EricMCornelius** [comment]
This is absolutely not in the spirit of semver - which is supposed to ensure no breaking changes are introduced for existing usage when making patches.

Your "API" absolutely is inclusive of any mandatory linked platform dependencies. Dynamic linking to hide those failures until runtime notwithstanding.

**Comment by bnoordhuis** [comment]
"In the spirit of" is one tremendous slippery slope. The fact is we follow semver. If semver had an explicit requirement of supporting platforms in perpetuity, we wouldn't have adopted it.

(Please don't take this as an invitation for further debate because it's not. Go over to https://github.com/semver/semver/issues where plenty of people have litigated the issue of platform support before.)

**Comment by jwdonahue** [comment]
For whatever it's worth it's worth here:

- [semever/semver#716: issuecomment-996252955](https://github.com/semver/semver/issues/716#issuecomment-996252955)
- [semver/semver#716: issuecomment-996291979](https://github.com/semver/semver/issues/716#issuecomment-996291979)

**Comment by introspection3** [comment]
bug

**Comment by introspection3** [comment]
> 微软去年放弃了对 Windows 7 的支持，所以我们也这样做了：#2821

your action is not good,you should maintain a good lib ,not follow microsoft

**Comment by gresolio** [comment]
As far as I understand, the following positions are the reason for the inability to solve this issue:
* Desire to drop Win7 support in order [to use new Win8 APIs](https://github.com/libuv/libuv/issues/3260#issuecomment-890938947).
* Lack of funding, ["Are you paying us?"](https://github.com/libuv/libuv/issues/3260#issuecomment-914984293).
* The introduced patch [#3285](https://github.com/libuv/libuv/pull/3285) "doesn't meet that bar": too long, non trivial, obtrusive.

The reasoning behind [bumping the minimum supported version to windows 8](https://github.com/libuv/libuv/pull/2821) looks strange. You may doubt the usefulness of Win7, but people are lazy by nature, they don't simply come to say thank you when everything works, but definitely going to annoy if something breaks.

About funding: if there was an official way to support you financially (e.g. donations, Patreon), many people including myself, would support you. Clearly defined funding objectives, for example "Win7 support", surely resonate in the hearts of people.

I think the patch [#3285](https://github.com/libuv/libuv/pull/3285) by kampeador is good enough. Even if there are some minor issues in "uv__gethostnamew_nt60" implementation, it will only affect users of the Win7 platform, which shouldn't bother you anymore:
```
  sGetHostNameW gethostnamew =
      pGetHostNameW == NULL ? uv__gethostnamew_nt60 : pGetHostNameW;
```
Why not just provide an opportunity for the community to help resolve Win7 issues?

**Comment by nicolasnoble** [comment]
What's pretty interesting to me with this issue still being opened and commented on is that #3340 basically fixed it, by just making the offending API not work on Windows 7, instead of crashing the whole software, or preventing compilation.

Why isn't #3340 good enough to keep libuv running on Windows 7? Only a minor API function won't work right on this os, with the correct error code, but otherwise everything else will be a-okay.

**Comment by Spudz76** [comment]
Because some software uses that call and expect it to work, and it was previously usable that way.

Also there is almost no way to figure out which software.  But I bet a lot of the very sudden Win7 deprecations widely across the industry is because of this change, and it's probably already too late.

At some point when your library is used by so many other things, it becomes a utility, and should not break backward compatibility even if you have good reasons.

**Comment by dr-dba** [comment]
The existing Windows 7 support should not be dropped, if the project claim to be "Cross-platform".
What is the point of "cross-platform" otherwise?
Ridiculous. It worked good, it still works good
just separate platform-specifics from non-platform specifics, as you should do anyway
I will not invest not in LibUV and not in any other project relying on it

**Comment by jwdonahue** [comment]
@dr-dba According to the [SUPPORTED_PLATFORMS list](https://github.com/libuv/libuv/blob/v1.x/SUPPORTED_PLATFORMS.md), LibUV has varying levels of support across 12 platforms. The fact that they dropped an antique Windows version from the list does not negate that fact.

@Spudz76  DOS, OS2 and WFW were all widely used enough to be considered "utilities", certainly more widely used than LubUV, yet none of those are supported anymore either. Applying your logic, we should never have banned the manufacture of lead pipes (tens of millions of homes depend on them for water around world!). 

It was Microsoft's planned and well publicized Win7 EOL date of 2020-01-14 that drove most of the deprecations across the industry (hardly sudden). You still have the option to fork the code base and continue to use it as you see fit.

**Comment by gresolio** [comment]
@jwdonahue Your logic is great, unfortunately the reasoning behind the drop is just strange and artificial.
See [my previous comment](https://github.com/libuv/libuv/issues/3260#issuecomment-1008421571) where all the official reasons are listed.

Forking is not always the best solution, it only fragments the community, funding the main project is a good alternative.

**Comment by stanleyxu2005** [comment]
> The existing Windows 7 support should not be dropped, if the project claim to be "Cross-platform". What is the point of "cross-platform" otherwise? Ridiculous. It worked good, it still works good just separate platform-specifics from non-platform specifics, as you should do anyway I will not invest not in LibUV and not in any other project relying on it

It's a very intestring discussion on the strategy of dropping support EOL platforms. On one hand, I don't see anything wrong that libuv maintainer is doing. On the other hand, the Windows 7 support removal does have a huge impact to it's downstream eco-system. Leveraging the impact, the Windows 7 support removal requires a techincal explaination.

@kampeador as you have already contributed a PR to add support for Windows 7. May I ask, if you are aware, that libuv 1.4.2 (or higher) is using any API that Windows 7 does not provide? If new libuv uses some new API that Windows 7 not provides, then it makes sense to drop the Windows 7 support.

**Comment by flawrist** [comment]
We need a hero to compile libuv for Windows 7.

**Comment by cjihrig** [comment]
> We need a hero to compile libuv for Windows 7.

Be the change you want to see in the world.
